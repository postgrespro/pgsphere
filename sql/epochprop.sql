SET extra_float_digits = 2;

SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               546.9759,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               -100) AS tp) AS q;

SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               0,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               -100) AS tp) AS q;

SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               NULL,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               -100) AS tp) AS q;

SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               23,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), NULL,
               20) AS tp) AS q;

SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               23,
               NULL, RADIANS(10362/3.6e6), -110,
               120) AS tp) AS q;

SELECT epoch_prop(NULL,
               23,
               0.01 , RADIANS(10362/3.6e6), -110,
               120);

SELECT epoch_prop_pos(spoint(radians(269.45207695), radians(4.693364966)), 
               23,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               20) AS tp;

SELECT epoch_prop_pos(spoint(radians(269.45207695), radians(4.693364966)), 
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6),
               20) AS tp;
