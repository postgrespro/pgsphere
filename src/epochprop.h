/* Definitions for handling proper motions */

#include <postgres.h>


Datum		epoch_prop(PG_FUNCTION_ARGS);


/*
 * a cartesian point; this is like geo_decl's point, but you can't
 * have both geo_decls and pg_sphere right now (both define a type Point,
 * not to mention they have different ideas on EPSILON.
 */
typedef struct s_cpoint
{
	double		x,
				y;
} CPoint;

typedef struct s_phasevec
{
	SPoint		pos;			/* Position as an SPoint */
	double		pm[2];			/* Proper motion long/lat in rad/year, PM in
								 * longitude has cos(lat) applied */
	double		parallax;		/* in rad */
	double		rv;				/* radial velocity in km/s */
	int			parallax_valid; /* 1 if the parallax really is a NULL */
} phasevec;
