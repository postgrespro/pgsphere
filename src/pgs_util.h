#ifndef __PGS_UTIL_H__
#define __PGS_UTIL_H__

#define PI			3.14159265358979323846		/* pi */
#define PIH			1.57079632679489661923		/* pi/2 */
#define PID			6.2831853071795864769		/* 2*pi */
#define RADIANS		57.295779513082320877		/* 180/pi */
#define PI_EPS		4.4408920985006261617e-16	/* 2 ** -51 */

#define Sqr(a)	 	( (a) * (a) )		/* square function as macro */

#ifdef EPSILON
#undef EPSILON
#endif
#define EPSILON		1.0E-09				/* precision of floating point values */

/* convert pg_sphere theta [pi/2 .. -pi/2] to healpix theta [0 .. pi] ([north .. south pole]) */
static inline double
conv_theta(double x)
{
	double y = PIH - x;
	if (fabs(x) < PI_EPS / 2)
		return PIH;
	if (fabs(y) < PI_EPS / 2)
		return 0;
	return y;
}

#endif
