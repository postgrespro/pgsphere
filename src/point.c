#include "point.h"

/* This file contains definitions for spherical point functions. */

PG_FUNCTION_INFO_V1(spherepoint_in);
PG_FUNCTION_INFO_V1(spherepoint_from_long_lat);
PG_FUNCTION_INFO_V1(spherepoint_distance);
PG_FUNCTION_INFO_V1(spherepoint_long);
PG_FUNCTION_INFO_V1(spherepoint_lat);
PG_FUNCTION_INFO_V1(spherepoint_x);
PG_FUNCTION_INFO_V1(spherepoint_y);
PG_FUNCTION_INFO_V1(spherepoint_z);
PG_FUNCTION_INFO_V1(spherepoint_xyz);
PG_FUNCTION_INFO_V1(spherepoint_equal);
PG_FUNCTION_INFO_V1(centroid);

bool
spoint_eq(const SPoint *p1, const SPoint *p2)
{
	Vector3D	a,
				b;

	spoint_vector3d(&a, p1);
	spoint_vector3d(&b, p2);
	return (vector3d_eq(&a, &b));
}

void
spoint_check(SPoint *spoint)
{
	bool lat_is_neg = (spoint->lat < 0) ? true : false;

	if (spoint->lng < 0 || spoint->lng > PID)
		spoint->lng = spoint->lng - floor(spoint->lng / (PID)) * PID;
	if (spoint->lat < -PIH || spoint->lat > PIH)
		spoint->lat = spoint->lat - floor(spoint->lat / (PID)) * PID;
	if (spoint->lng < 0.0)
	{
		spoint->lng += (PID);
	}
	if (spoint->lat > PI)
	{
		spoint->lat -= (2 * PI);
	}
	if (spoint->lat > PIH)
	{
		spoint->lat = (PI - spoint->lat);
		spoint->lng += ((spoint->lng < PI) ? (PI) : (-PI));
	}
	if (spoint->lat < -PIH)
	{
		spoint->lat = (-PI - spoint->lat);
		spoint->lng += ((spoint->lng < PI) ? (PI) : (-PI));
	}
	if (FPeq(spoint->lat, PIH) && lat_is_neg)
		spoint->lat = -PIH;

	if (FPeq(spoint->lng, PID))
	{
		spoint->lng = 0.0;
	}
	if (FPzero(spoint->lng))
	{
		spoint->lng = 0.0;
	}
	if (FPzero(spoint->lat))
	{
		spoint->lat = 0.0;
	}
}

void
vector3d_spoint(SPoint *p, const Vector3D *v)
{
	double rho = sqrt((v->x) * (v->x) + (v->y) * (v->y));

	if (0.0 == rho)
	{
		if (FPzero(v->z))
		{
			p->lat = 0.0;
		}
		else if (v->z > 0)
		{
			p->lat = PIH;
		}
		else if (v->z < 0)
		{
			p->lat = -PIH;
		}
	}
	else
	{
		p->lat = atan(v->z / rho);
	}

	p->lng = atan2(v->y, v->x);
	if (FPzero(p->lng))
	{
		p->lng = 0.0;
	}
	else if (p->lng < 0.0)
	{
		p->lng += PID;
	}
}

void
spoint_vector3d(Vector3D *v, const SPoint *p)
{
	v->x = cos(p->lng) * cos(p->lat);
	v->y = sin(p->lng) * cos(p->lat);
	v->z = sin(p->lat);
}

Datum
spherepoint_in(PG_FUNCTION_ARGS)
{
	SPoint	   *sp = (SPoint *) palloc(sizeof(SPoint));
	char	   *c = PG_GETARG_CSTRING(0);
	double		lng,
				lat;

	void		sphere_yyparse(void);

	init_buffer(c);
	sphere_yyparse();
	if (get_point(&lng, &lat))
	{
		sp->lng = lng;
		sp->lat = lat;
		spoint_check(sp);
	}
	else
	{
		reset_buffer();
		pfree(sp);
		sp = NULL;
		elog(ERROR, "spherepoint_in: parse error");
	}
	reset_buffer();
	PG_RETURN_POINTER(sp);
}


Datum
spherepoint_from_long_lat(PG_FUNCTION_ARGS)
{
	SPoint *p = (SPoint *) palloc(sizeof(SPoint));

	p->lng = PG_GETARG_FLOAT8(0);
	p->lat = PG_GETARG_FLOAT8(1);
	spoint_check(p);
	PG_RETURN_POINTER(p);
}

static double
norm2(double a, double b)
{
	return sqrt(a * a + b * b);
}

float8
spoint_dist(const SPoint *p1, const SPoint *p2)
{
	float8	dl = p1->lng - p2->lng;
	/* use Vincenty's formula for the inverse geodesic problem on the sphere */
	float8	f = atan2(norm2(cos(p2->lat) * sin(dl),
							cos(p1->lat) * sin(p2->lat)
								- sin(p1->lat) * cos(p2->lat) * cos(dl)),
						sin(p1->lat) * sin(p2->lat)
							+ cos(p1->lat) * cos(p2->lat) * cos(dl));
	if (FPzero(f))
	{
		return 0.0;
	}
	else
	{
		return f;
	}
}

Datum
spherepoint_distance(PG_FUNCTION_ARGS)
{
	SPoint	   *p1 = (SPoint *) PG_GETARG_POINTER(0);
	SPoint	   *p2 = (SPoint *) PG_GETARG_POINTER(1);

	PG_RETURN_FLOAT8(spoint_dist(p1, p2));

}

Datum
spherepoint_long(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(p->lng);
}

Datum
spherepoint_lat(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);

	PG_RETURN_FLOAT8(p->lat);
}

Datum
spherepoint_x(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Vector3D	v;

	spoint_vector3d(&v, p);
	PG_RETURN_FLOAT8(v.x);
}

Datum
spherepoint_y(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Vector3D	v;

	spoint_vector3d(&v, p);
	PG_RETURN_FLOAT8(v.y);
}

Datum
spherepoint_z(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Vector3D	v;

	spoint_vector3d(&v, p);
	PG_RETURN_FLOAT8(v.z);
}

Datum
spherepoint_xyz(PG_FUNCTION_ARGS)
{
	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
	Datum		dret[3];
	ArrayType  *result;
	Vector3D	v;

	spoint_vector3d(&v, p);
	dret[0] = Float8GetDatumFast(v.x);
	dret[1] = Float8GetDatumFast(v.y);
	dret[2] = Float8GetDatumFast(v.z);
#ifdef USE_FLOAT8_BYVAL
	result = construct_array(dret, 3, FLOAT8OID, sizeof(float8), true, 'd');
#else
	result = construct_array(dret, 3, FLOAT8OID, sizeof(float8), false, 'd');
#endif
	PG_RETURN_ARRAYTYPE_P(result);
}

Datum
spherepoint_equal(PG_FUNCTION_ARGS)
{
	SPoint	   *p1 = (SPoint *) PG_GETARG_POINTER(0);
	SPoint	   *p2 = (SPoint *) PG_GETARG_POINTER(1);

	PG_RETURN_BOOL(spoint_eq(p1, p2));
}

static SPoint * spherepoint_from_vector3d(Vector3D v){
	SPoint* p = (SPoint *) palloc(sizeof(SPoint));
	p->lat = asin(v.z / sqrt(pow(v.x, 2) + pow(v.y, 2)  + pow(v.z, 2)));
	p->lng = atan2(v.y, v.x);
	return p;
}

Datum centroid(PG_FUNCTION_ARGS) {
	int i;
	SPoint * p;
	SPoint current_point;
	Vector3D	v;
	Vector3D	point_coords;
	ArrayType *dots_vector = PG_GETARG_ARRAYTYPE_P(0);
	int num_elements = ArrayGetNItems(ARR_NDIM(dots_vector), ARR_DIMS(dots_vector));
	SPoint *array_data = (SPoint *) ARR_DATA_PTR(dots_vector);

	for (i = 0; i < num_elements; i++) {
		current_point = array_data[i];
		spoint_vector3d(&v, &current_point);
		point_coords.x+=v.x;
		point_coords.y+=v.y;
		point_coords.z+=v.z;
	}

	point_coords.x /= num_elements;
	point_coords.y /= num_elements;
	point_coords.z /= num_elements;

	p = spherepoint_from_vector3d(point_coords);

	spoint_check(p);

	PG_RETURN_POINTER(p);
}
