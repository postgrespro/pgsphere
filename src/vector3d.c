#include "vector3d.h"

/* Vector functions */

bool
vector3d_eq(const Vector3D *v1, const Vector3D *v2)
{
	return (FPeq(v1->x, v2->x) &&
			FPeq(v1->y, v2->y) &&
			FPeq(v1->z, v2->z));
}

void
vector3d_cross(Vector3D *out, const Vector3D *v1, const Vector3D *v2)
{
	out->x = v1->y * v2->z - v1->z * v2->y;
	out->y = v1->z * v2->x - v1->x * v2->z;
	out->z = v1->x * v2->y - v1->y * v2->x;
}

void
vector3d_addwithscalar(Vector3D *result, double scalar, const Vector3D *delta)
{
	result->x = result->x + delta->x * scalar;
	result->y = result->y + delta->y * scalar;
	result->z = result->z + delta->z * scalar;
}

float8
vector3d_scalar(Vector3D *v1, Vector3D *v2)
{
	float8		out = 0;

	out += v1->x * v2->x;
	out += v1->y * v2->y;
	out += v1->z * v2->z;
	return (out);
}

float8
vector3d_length(const Vector3D *v)
{
	return sqrt(Sqr(v->x) + Sqr(v->y) + Sqr(v->z));
}
