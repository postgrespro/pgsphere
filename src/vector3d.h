#ifndef __PGS_VECTOR3D_H__
#define __PGS_VECTOR3D_H__

#include "pg_sphere.h"

/* Vector declarations */

/*
 * The definition of a three dimensional vector data structure.
 */
typedef struct
{
	float8	x;	/* x (-1.0 .. 1.0) */
	float8	y;	/* y (-1.0 .. 1.0) */
	float8	z;	/* z (-1.0 .. 1.0) */
} Vector3D;


/*
 * Calculate the cross product of two vectors. Puts
 * cross product of v1 and v2 into out and returns it.
 */
void	vector3d_cross(Vector3D *out, const Vector3D *v1, const Vector3D *v2);

/*
 * Checks equality of two vectors.
 */
bool	vector3d_eq(const Vector3D *a, const Vector3D *b);

/*
 * Calculate the scalar product of two vectors.
 */
float8	vector3d_scalar(Vector3D *v1, Vector3D *v2);

/*
 * Calculate the length of a vector.
 */
float8	vector3d_length(const Vector3D *v);

#endif
