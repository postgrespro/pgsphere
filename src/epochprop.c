/* Handling of (conventional) proper motions.

This code is largely based on a FORTRAN function written by Lennart Lindegren
(Lund Obs) in 1995 that implements the procedure described in The Hipparcos
and Tycho Catalogues (ESA SP-1200), Volume 1, Section 1.5.5, 'Epoch
Transformation: Rigorous Treatment'; cf.
<https://www.cosmos.esa.int/documents/532822/552851/vol1_all.pdf/99adf6e3-6893-4824-8fc2-8d3c9cbba2b5>.
*/

#include <math.h>
#include "pgs_util.h"

#include "point.h"
#include "epochprop.h"
#include "vector3d.h"

PG_FUNCTION_INFO_V1(epoch_prop);

/* Astronomical unit in kilometers */
#define AU 1.495978701e8

/* Julian year in seconds */
#define J_YEAR (365.25*86400)

/* A_nu as per ESA/SP-1200 */
#define A_NU (AU/(J_YEAR))

/* Following SOFA, we use 1e-7 arcsec as minimal parallax
	("celestial sphere"); parallax=0 exactly means "infinite distance", which
	leads to all sorts for problems; our parallaxes come in in mas, so: */
#define PX_MIN 1e-7*1000

/* propagate an object at a phase vector over a time difference of delta_t,
stuffing an updated phase vector in result.

This does not propagate errors.
*/
static void
propagate_phasevec(const phasevec *pv,
				   const double delta_t,
				   phasevec *result)
{

	double		distance_factor,
				mu0abs,
				zeta0,
				parallax;

	Vector3D	p0,
				r0,
				q0;
	Vector3D	mu0,
				pprime,
				qprime,
				mu,
				muprime,
				u,
				uprime;

	/*
	 * for very small or null parallaxes, our algorithm breaks; avoid that
	 * and, if we did emergency measures, do not talk about parallax and
	 * radial velocity in the output
	 */
	if (pv->parallax_valid)
	{
		parallax = pv->parallax;
	}
	else
	{
		parallax = PX_MIN;
	}
	result->parallax_valid = pv->parallax_valid;

	/* compute the normal triad as Vector3D-s, eq. (1.2.15) */
	spoint_vector3d(&r0, &(pv->pos));

	p0.x = -sin(pv->pos.lng);
	p0.y = cos(pv->pos.lng);
	p0.z = 0;

	q0.x = -sin(pv->pos.lat) * cos(pv->pos.lng);
	q0.y = -sin(pv->pos.lat) * sin(pv->pos.lng);
	q0.z = cos(pv->pos.lat);

	/* the original proper motion vector */
	mu0.x = mu0.y = mu0.z = 0;
	vector3d_addwithscalar(&mu0, pv->pm[0], &p0);
	vector3d_addwithscalar(&mu0, pv->pm[1], &q0);
	mu0abs = vector3d_length(&mu0);

	/*
	 * radial velocity in mas/yr ("change of parallax per year").  eq.
	 * (1.5.24) We're transforming this to rad/yr so the units work out below
	 */
	zeta0 = (pv->rv * parallax / A_NU) / 3.6e6 / RADIANS;
	/* distance factor eq. (1.5.25) */
	distance_factor = 1 / sqrt(1
							   + 2 * zeta0 * delta_t
							   + (mu0abs * mu0abs + zeta0 * zeta0) * delta_t * delta_t);

	/* the propagated proper motion vector, eq. (1.5.28) */
	muprime.x = muprime.y = muprime.z = 0;
	vector3d_addwithscalar(&muprime, (1 + zeta0 * delta_t), &mu0);
	vector3d_addwithscalar(&muprime, -mu0abs * mu0abs * delta_t, &r0);
	mu.x = mu.y = mu.z = 0;
	vector3d_addwithscalar(&mu, pow(distance_factor, 3), &muprime);

	/* parallax, eq. (1.5.27) */
	result->parallax = distance_factor * parallax;
	/* zeta/rv, eq. (1.5.29); go back from rad to mas, too */
	result->rv = (zeta0 + (mu0abs * mu0abs + zeta0 * zeta0) * delta_t)
		* distance_factor * distance_factor
		* 3.6e6 * RADIANS
		* A_NU / result->parallax;

	/* propagated position, eq. (1.5.26) */
	uprime.x = uprime.y = uprime.z = 0;
	vector3d_addwithscalar(&uprime, (1 + zeta0 * delta_t), &r0);
	vector3d_addwithscalar(&uprime, delta_t, &mu0);
	u.x = u.y = u.z = 0;
	vector3d_addwithscalar(&u, distance_factor, &uprime);
	vector3d_spoint(&(result->pos), &u);

	/* compute a new triad for the propagated position, eq (1.5.15) */
	pprime.x = -sin(result->pos.lng);
	pprime.y = cos(result->pos.lng);
	pprime.z = 0;
	qprime.x = -sin(result->pos.lat) * cos(result->pos.lng);
	qprime.y = -sin(result->pos.lat) * sin(result->pos.lng);
	qprime.z = cos(result->pos.lat);

	/* use it to compute the proper motions, eq. (1.5.32) */
	result->pm[0] = vector3d_scalar(&pprime, &mu);
	result->pm[1] = vector3d_scalar(&qprime, &mu);
}


/*
	Propagate a position with proper motions and optionally parallax
	and radial velocity.

	Arguments: pos0 (spoint), pm_long, pm_lat (in rad/yr)
	par (parallax, mas), rv (in km/s), delta_t (in years)

	This returns a 6-array of lat, long (in rad), parallax (in mas)
	pmlat, pmlong (in rad/yr), rv (in km/s).
*/
Datum
epoch_prop(PG_FUNCTION_ARGS)
{
	double		delta_t;
	phasevec	input,
				output;
	ArrayType  *result;
	Datum		retvals[6];

	if (PG_ARGISNULL(0))
	{
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("NULL position not supported in epoch propagation")));
	}
	memcpy(&(input.pos), (void *) PG_GETARG_POINTER(0), sizeof(SPoint));
	if (PG_ARGISNULL(1))
	{
		input.parallax = 0;
	}
	else
	{
		input.parallax = PG_GETARG_FLOAT8(1);
	}
	input.parallax_valid = fabs(input.parallax) > PX_MIN;

	if (PG_ARGISNULL(2))
	{
		input.pm[0] = 0;
	}
	else
	{
		input.pm[0] = PG_GETARG_FLOAT8(2);
	}

	if (PG_ARGISNULL(3))
	{
		input.pm[1] = 0;
	}
	else
	{
		input.pm[1] = PG_GETARG_FLOAT8(3);
	}

	if (PG_ARGISNULL(4))
	{
		input.rv = 0;
	}
	else
	{
		input.rv = PG_GETARG_FLOAT8(4);
	}

	if (PG_ARGISNULL(5))
	{
		ereport(ERROR,
				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
				 errmsg("NULL delta t not supported in epoch propagation")));
	}
	delta_t = PG_GETARG_FLOAT8(5);

	propagate_phasevec(&input, delta_t, &output);

	/* change to internal units: rad, rad/yr, mas, and km/s */
	retvals[0] = Float8GetDatum(output.pos.lng);
	retvals[1] = Float8GetDatum(output.pos.lat);
	retvals[2] = Float8GetDatum(output.parallax);
	retvals[3] = Float8GetDatum(output.pm[0]);
	retvals[4] = Float8GetDatum(output.pm[1]);
	retvals[5] = Float8GetDatum(output.rv);

	{
		bool		isnull[6] = {0, 0, 0, 0, 0, 0};
		int			lower_bounds[1] = {1};
		int			dims[1] = {6};
#ifdef USE_FLOAT8_BYVAL
		bool		embyval = true;
#else
		bool		embyval = false;
#endif

		if (!output.parallax_valid)
		{
			/* invalidate parallax and rv */
			isnull[2] = 1;
			isnull[5] = 1;
		}

		result = construct_md_array(retvals, isnull, 1, dims, lower_bounds,
									FLOAT8OID, sizeof(float8), embyval, 'd');
	}
	PG_RETURN_ARRAYTYPE_P(result);
}
