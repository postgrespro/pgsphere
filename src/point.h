#ifndef __PGS_POINT_H__
#define __PGS_POINT_H__

#include "vector3d.h"
#include "sbuffer.h"

/* This file contains declarations for spherical point and functions. */

/*
 * The data structure definition of a spherical point.
 */
typedef struct
{
	float8		lng;			/* longitude value in radians */
	float8		lat;			/* latitude value in radians */
} SPoint;

extern Oid	get_spoint_type_oid(void);

/*
 * Calculate the distance between two spherical points in radians.
 */

extern float8 spoint_dist(const SPoint *p1, const SPoint *p2);

/*
 * Check whether two points are equal.
 */

extern bool spoint_eq(const SPoint *p1, const SPoint *p2);

/*
 * Check the longitude and latitude values of a spherical point.
 */

extern void spoint_check(SPoint *spoint);

/*
 * Transforms a 3d vector into a spherical point.
 */

extern void vector3d_spoint(SPoint *p, const Vector3D *v);

/*
 * Transforms a spherical point into a 3d vector.
 */

extern void spoint_vector3d(Vector3D *v, const SPoint *p);

/*
 * Take the input and store it as a spherical point.
 */

extern Datum spherepoint_in(PG_FUNCTION_ARGS);

/*
 * Create spherical point from lat, lng and store to first argument(pointer)
 */

extern void create_spherepoint_from_long_lat(SPoint *p, float8 lng, float8 lat);

/*
 * Create a spherical point from longitude and latitude both in radians.
 */

extern Datum spherepoint_from_long_lat(PG_FUNCTION_ARGS);

/*
 * Create a spherical point from longitude and latitude both in degrees.
 */

extern Datum spherepoint_from_long_lat_deg(PG_FUNCTION_ARGS);

/*
 * Calculate the distance between two spherical points.
 */
extern Datum spherepoint_distance(PG_FUNCTION_ARGS);

/*
 * Longitude of a spherical point.
 */
extern Datum spherepoint_long(PG_FUNCTION_ARGS);

/*
 * Latitude of a spherical point.
 */
extern Datum spherepoint_lat(PG_FUNCTION_ARGS);

/*
 * Cartesian x-value of a spherical point.
 */
extern Datum spherepoint_x(PG_FUNCTION_ARGS);

/*
 * Cartesian y-value of a spherical point.
 */
extern Datum spherepoint_y(PG_FUNCTION_ARGS);

/*
 * Cartesian z-value of a spherical point.
 */
extern Datum spherepoint_z(PG_FUNCTION_ARGS);

/*
 * Cartesian values of a spherical point as an array.
 */
extern Datum spherepoint_xyz(PG_FUNCTION_ARGS);

/*
 * Check whether two points are equal.
 */
extern Datum spherepoint_equal(PG_FUNCTION_ARGS);

/*
 * Compute a 32-bit hash value of a point.
 */
extern Datum spherepoint_hash32(PG_FUNCTION_ARGS);

#endif
