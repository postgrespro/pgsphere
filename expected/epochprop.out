SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               546.9759,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               -100) AS tp) AS q;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 9:        SELECT epoch_prop(spoint(radians(269.45207695), radia...
                                 ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               0,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               -100) AS tp) AS q;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 9:        SELECT epoch_prop(spoint(radians(269.45207695), radia...
                                 ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               NULL,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               -100) AS tp) AS q;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 9:        SELECT epoch_prop(spoint(radians(269.45207695), radia...
                                 ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               23,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), NULL,
               20) AS tp) AS q;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 9:        SELECT epoch_prop(spoint(radians(269.45207695), radia...
                                 ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT 
       to_char(DEGREES(tp[1]), '999D9999999999'),
       to_char(DEGREES(tp[2]), '999D9999999999'),
       to_char(tp[3], '999D999'),
       to_char(DEGREES(tp[4])*3.6e6, '999D999'),
       to_char(DEGREES(tp[5])*3.6e6, '99999D999'),
       to_char(tp[6], '999D999')
FROM (
       SELECT epoch_prop(spoint(radians(269.45207695), radians(4.693364966)), 
               23,
               NULL, RADIANS(10362/3.6e6), -110,
               120) AS tp) AS q;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 9:        SELECT epoch_prop(spoint(radians(269.45207695), radia...
                                 ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT epoch_prop(NULL,
               23,
               0.01 , RADIANS(10362/3.6e6), -110,
               120);
ERROR:  function epoch_prop(unknown, integer, numeric, double precision, integer, integer) does not exist
LINE 1: SELECT epoch_prop(NULL,
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT epoch_prop_pos(spoint(radians(269.45207695), radians(4.693364966)), 
               23,
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6), -110,
               20) AS tp;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 1: SELECT epoch_prop_pos(spoint(radians(269.45207695), radians(...
                              ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
SELECT epoch_prop_pos(spoint(radians(269.45207695), radians(4.693364966)), 
               RADIANS(-801.551/3.6e6), RADIANS(10362/3.6e6),
               20) AS tp;
ERROR:  function spoint(double precision, double precision) does not exist
LINE 1: SELECT epoch_prop_pos(spoint(radians(269.45207695), radians(...
                              ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
